Firestore/core/src/firebase/firestore/remote/connectivity_monitor_apple.mm

```diff
@@ -0,0 +1,130 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/connectivity_monitor.h"
+
+#import <Foundation/Foundation.h>
+#import <SystemConfiguration/SystemConfiguration.h>
+#import <netinet/in.h>
+#include <memory>
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "absl/memory/memory.h"
+#include "dispatch/dispatch.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+namespace {
+
+using NetworkStatus = ConnectivityMonitor::NetworkStatus;
+using util::AsyncQueue;
+
+NetworkStatus ToNetworkStatus(SCNetworkReachabilityFlags flags) {
+  if (!(flags & kSCNetworkReachabilityFlagsReachable)) {
+    return NetworkStatus::Unreachable;
+  }
+  if (flags & kSCNetworkReachabilityFlagsConnectionRequired) {
+    return NetworkStatus::Unreachable;
+  }
+
+#if TARGET_OS_IPHONE
+  if (flags & kSCNetworkReachabilityFlagsIsWWAN) {
+    return NetworkStatus::ReachableViaCellular;
+  }
+#endif
+  return NetworkStatus::Reachable;
+}
+
+SCNetworkReachabilityRef CreateReachability() {
+  // Pseudoaddress that monitors internet reachability in general.
+  sockaddr_in any_connection_addr{};
+  any_connection_addr.sin_len = sizeof(any_connection_addr);
+  any_connection_addr.sin_family = AF_INET;
+  return SCNetworkReachabilityCreateWithAddress(
+      nullptr, reinterpret_cast<sockaddr*>(&any_connection_addr));
+}
+
+}  // namespace
+
+/**
+ * Implementation of `ConnectivityMonitor` based on `SCNetworkReachability`
+ * (iOS/MacOS).
+ */
+class ConnectivityMonitorApple : public ConnectivityMonitor {
+ public:
+  explicit ConnectivityMonitorApple(AsyncQueue* worker_queue)
+      : ConnectivityMonitor{worker_queue}, reachability_{CreateReachability()} {
+    SCNetworkReachabilityFlags flags;
+    if (SCNetworkReachabilityGetFlags(reachability_, &flags)) {
+      SetInitialStatus(ToNetworkStatus(flags));
+    }
+
+    auto on_reachability_changed = [](SCNetworkReachabilityRef /*unused*/,
+                                      SCNetworkReachabilityFlags flags,
+                                      void* raw_this) {
+      HARD_ASSERT(raw_this, "Received a null pointer as context");
+      static_cast<ConnectivityMonitorApple*>(raw_this)->OnChange(flags);
+    };
+
+    SCNetworkReachabilityContext context{};
+    context.info = this;
+    bool success = SCNetworkReachabilitySetCallback(
+        reachability_, on_reachability_changed, &context);
+    if (!success) {
+      LOG_DEBUG("Couldn't set reachability callback");
+      return;
+    }
+
+    auto queue =
+        dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0);
+    success = SCNetworkReachabilitySetDispatchQueue(reachability_, queue);
```

Comment:
Why not set the dispatch queue to our worker queue and then call [`ExecuteBlocking`](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/util/async_queue.h#L128) on callback?

Also, shouldn't we set the queue before we set the callback to avoid a first callback on some other default queue?

Reply:
ExecuteBlocking exists essentially for any API that allows us to set a dispatch queue on which to get callbacks.

Far from undesirable, this makes it possible to avoid a double-bounce on every callback. For example, without it, as implemented you get a callback on the main queue which then enqueues onto our worker. If use use ExecuteBlocking instead you can instruct SCNetworkReachability to call back on the worker directly (and then we just reestablish our invariants ourselves.

Reply:
Done.

`AsyncQueue` doesn't know anything about libdispatch. I can see two ways around it:
* add a method to the queue which returns the underlying executor, which can then be cast to `ExecutorLibdispatch`;
* add a `dispatch_queue` getter to `AsyncQueue` and wrap it inside a `ifdef APPLE` guard.

Implemented the former alternative, let me know what you think.

Also, this necessitated changing the file type back to Objective-C++. I've recently ran into trouble including `executor_libdispatch.h` from both C++ and Objective-C files because `dispatch_queue_t` got defined to different types, leading to link errors.